app/src/main/java/com/x8bit/bitwarden/data/autofill/parser/AutofillParserImpl.kt

```diff
@@ -139,7 +144,50 @@ class AutofillParserImpl(
 private fun AssistStructure.traverse(): List<ViewNodeTraversalData> =
     (0 until windowNodeCount)
         .map { getWindowNodeAt(it) }
-        .mapNotNull { windowNode -> windowNode.rootViewNode?.traverse() }
+        .mapNotNull { windowNode ->
+            windowNode
+                .rootViewNode
+                ?.traverse()
+                ?.updateForMissingUsernameFields()
+        }
+
+/**
+ * This helper function updates the [ViewNodeTraversalData] is necessary for missing username
+ * fields that could have been missed. If the current `ViewNodeTraversalData` contains password
+ * fields but no username fields, we check to see if there are any unused fields directly above
+ * the password fields and we assume that it those are the missing username fields.
+ */
+private fun ViewNodeTraversalData.updateForMissingUsernameFields(): ViewNodeTraversalData {
+    val passwordPositions = this.autofillViews.mapIndexedNotNull { index, autofillView ->
+        (autofillView as? AutofillView.Login.Password)?.let { index }
+    }
+    return if (passwordPositions.any() &&
+        this.autofillViews.none { it is AutofillView.Login.Username }
+    ) {
+        val updatedAutofillViews = autofillViews.mapIndexed { index, autofillView ->
+            if (autofillView is AutofillView.Unused && passwordPositions.contains(index + 1)) {
+                AutofillView.Login.Username(data = autofillView.data)
+            } else {
+                autofillView
+            }
+        }
+        val previousUnusedIds = autofillViews
+            .filterIsInstance<AutofillView.Unused>()
+            .map { it.data.autofillId }
+            .toSet()
+        val currentUnusedIds = updatedAutofillViews
+            .filterIsInstance<AutofillView.Unused>()
+            .map { it.data.autofillId }
+            .toSet()
+        val unignoredAutofillIds = previousUnusedIds - currentUnusedIds
+        this.copy(
+            autofillViews = updatedAutofillViews,
+            ignoreAutofillIds = this.ignoreAutofillIds.minus(unignoredAutofillIds),
+        )
+    } else {
+        this
```

Comment:
Can we either use an early return for this or at least add a comment here making clear what this represents? Like:

```
// No unused username fields available above password fields,
// so nothing to modify.
this
```


By the time you get down here it's easy to have lost the thread of what the check is.